File: src/main/java/com/dik/models/Assembly.java
package com.dik.models;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by billa on 2016-12-01.
 */
public class Assembly {
    private String name;
    private List<Operation> operations;
    private List<Operation> allocatedOperations;

    public Assembly(String name, List<Operation> operations) {
        this.name = name;
        this.operations = operations;
        allocatedOperations = new ArrayList<>();
    }

    public boolean allocateOperation(Operation operation){
        if(operations.contains(operation)){
            allocatedOperations.add(operation);
            return true;
        }
        return false;
    }

    public List<Operation> getAllocatedOperations() {
        return allocatedOperations;
    }

    public String getName() {

        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<Operation> getOperations() {
        return operations;
    }

    public void setOperations(List<Operation> operations) {
        this.operations = operations;
    }
}

File: app/models/Assembly.scala
package models

import scala.collection.mutable

/**
  * Created by billa on 2016-12-15.
  */

sealed trait AssemblyOperationStatus{
  def text:String
}

case object BusyOperationStatus extends AssemblyOperationStatus{
  def text = "busy"
}
case object FreeOperationStatus extends AssemblyOperationStatus{
  def text = "free"
}

object AssemblyOperationStatus{
  def apply(text:String) ={
    text match {
      case "busy" => BusyOperationStatus
      case "free" => FreeOperationStatus
      case "q" => FreeOperationStatus
    }
  }
}

case class AssemblyOperation(operation:Operation,time:Int,status:AssemblyOperationStatus)

case class Assembly(id:Int,name:String,totalOperations:List[AssemblyOperation] =List[AssemblyOperation](),
                    allocatedOperations:List[AssemblyOperation]) {

//  def freeOperation(maybeOperation: Option[Operation]):Unit ={
//    maybeOperation match{
//      case Some(x) => {
//        allocatedOperations -= x
//      }
//      case None =>
//    }
//  }
//
//  def allocateOperation(operation: Operation): Boolean ={
//    synchronized {
//      totalOperations.filter(_.operation.equals(operation)).contains(operation) match {
//        case true =>
//          allocatedOperations += operation
//          true
//        case false => false
//      }
//    }

//  }
}
